### PR TITLE
Minor cleanup in deframer

### PR DESCRIPTION
**Pull request by carl-mastrangelo**
After looking through the deframer, I am pretty convinced it is correct.  I added some minor cleanups to things, but I think the logic is okay.  Tests should be written.

### CODE REVIEW

File: core/src/main/java/io/grpc/internal/MessageDeframer.java [hunk 1]
```diff
@@ -258,9 +258,11 @@ private void deliver() {
       stalled = !isDataAvailable();
 
       if (endOfStream) {
-        if (!isDataAvailable()) {
+        if (!stalled) {
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/MessageDeframer.java]: Uhh... doesn't this invert the logic? `!stalled == isDataAvailable()`

**Comment by ejona86** [comment]
What do you think the purpose of stalled is? Stalled's only purpose was exactly the case that was broken.

Consider this comment in `MessageDeframer`:
```java
       // We are stalled when there are no more bytes to process. This allows delivering errors as
       // soon as the buffered input has been consumed, independent of whether the application
       // has requested another message.
       stalled = !isDataAvailable();
```

And this one in `AbstractClientStream`:
```java
    // Determine if the deframer is stalled (i.e. currently has no complete messages to deliver).
    boolean deliveryStalled = isDeframerStalled();
```

If we deem the stalled behavior currently correct, wouldn't both of those comments be incorrect?

File: core/src/main/java/io/grpc/internal/MessageDeframer.java (hunk 1 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/internal/MessageDeframer.java]: Done.

**Comment by carl-mastrangelo** [comment]
From my interpretation, stalled means the deframer currently has a partial message. (header and all).  Thats why I think the isDataAvailable method is correct, because the partial message could be inside the unprocessed buffer or the nextFrame buffer.

**Comment by ejona86** [comment]
Stalled was supposed to be true if the deframer was not guaranteed to produce more output messages given the current input data (excluding application flow control). It was explicitly for error cases, where we wanted to "drain" the `MessageDeframer` and provide an error after that point.

More comments that don't seem to agree with your definition. This one in `MessageDeframer`:
```java
/**
 * Called when end-of-stream has not yet been reached but there are no complete messages
 * remaining to be delivered.
 */
void deliveryStalled();
```

And this comment in `AbstractStream` (for the only method called by an implementation of `deliveryStalled()` listener):
```java
/**
 * Handles the event that the deframer has no pending deliveries.
 */
protected abstract void inboundDeliveryPaused();
```

File: core/src/main/java/io/grpc/internal/MessageDeframer.java (hunk 1 above)
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/MessageDeframer.java]: The fix isn't correct either. You are inverting the meaning of stalled. The fix doesn't pass the tests just like the original commit.

**Comment by carl-mastrangelo** [comment]
AbstractStream I think is actually wrong, and thus the source of the bug.

**Comment by carl-mastrangelo** [comment]
PTAL.

Note that since DF starts stalled, and almost always ends stalled, the stalled listener will be called less.

File: core/src/main/java/io/grpc/internal/MessageDeframer.java [hunk 2]
```diff
@@ -183,7 +186,8 @@ public void deframe(ReadableBuffer data, boolean endOfStream) {
   }
 
   /**
-   * Indicates whether delivery is currently stalled, pending receipt of more data.
+   * Indicates whether delivery is currently stalled, pending receipt of more data.  This means
+   * that no data additional data can be delivered to the application.
```
Comment (madongfly) [inline core/src/main/java/io/grpc/internal/MessageDeframer.java]: "that no additional data ..."

File: core/src/main/java/io/grpc/internal/MessageDeframer.java [hunk 3]
```diff
@@ -153,11 +155,12 @@ public void request(int numMessages) {
   }
 
   /**
-   * Adds the given data to this deframer and attempts delivery to the sink.
+   * Adds the given data to this deframer and attempts delivery to the listener.
    *
    * @param data the raw data read from the remote endpoint. Must be non-null.
    * @param endOfStream if {@code true}, indicates that {@code data} is the end of the stream from
-   *        the remote endpoint.
+   *        the remote endpoint.  End of stream should not be used in the event of an error, such as
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/MessageDeframer.java]: s/error/transport error/ ? Or really any other clarification that makes it obvious we aren't talking about the grpc status code.

**Comment by ejona86** [comment]
@carl-mastrangelo, LGTM. Thank you!

**Comment by nmittler** [comment]
@carl-mastrangelo logic looks much better ... LGTM!

File: core/src/main/java/io/grpc/internal/MessageDeframer.java (hunk 2 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/internal/MessageDeframer.java]: Done.

File: core/src/main/java/io/grpc/internal/MessageDeframer.java (hunk 3 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/internal/MessageDeframer.java]: Done.

**Pull request closed without merging**